primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs

```diff
@@ -80,66 +86,40 @@ where
         merkle_proof: Rescue3AryMerklePathVar,
         merkle_root: Variable,
     ) -> Result<BoolVar, CircuitError> {
-        let root_var = self.compute_merkle_root(elem, &merkle_proof)?;
+        let root_var =
+            <PlonkCircuit<F> as MerkleTreeHelperGadget<RescueMerkleTree<F>>>::compute_merkle_root(
+                self,
+                elem,
+                &merkle_proof,
+            )?;
         self.is_equal(root_var, merkle_root)
     }
 
-    fn enforce_merkle_proof(
+    fn enforce_membership_proof(
         &mut self,
         elem: StandardLeafVar,
         merkle_proof: Rescue3AryMerklePathVar,
         expected_merkle_root: Variable,
     ) -> Result<(), CircuitError> {
-        let bool_val = self
-            .is_member(elem, merkle_proof, expected_merkle_root)
-            .unwrap();
+        let bool_val = <PlonkCircuit<F> as MerkleTreeGadget<RescueMerkleTree<F>>>::is_member(
+            self,
+            elem,
+            merkle_proof,
+            expected_merkle_root,
+        )
+        .unwrap();
         self.enforce_true(bool_val.into())
     }
 }
 
-#[derive(PartialEq, Eq, Clone, Debug)]
-struct MerkleNodeBooleanEncoding<F: RescueParameter> {
-    sibling1: NodeVal<F>,
-    sibling2: NodeVal<F>,
-    is_left_child: bool,
-    is_right_child: bool,
-}
-
-impl<F: RescueParameter> MerkleNodeBooleanEncoding<F> {
-    fn new(
-        sibling1: NodeVal<F>,
-        sibling2: NodeVal<F>,
-        is_left_child: bool,
-        is_right_child: bool,
-    ) -> Self {
-        MerkleNodeBooleanEncoding {
-            sibling1,
-            sibling2,
-            is_left_child,
-            is_right_child,
-        }
-    }
-}
-
-#[derive(PartialEq, Eq, Clone, Debug)]
-struct MembershipProofBooleanEncoding<F: RescueParameter> {
-    pub nodes: Vec<MerkleNodeBooleanEncoding<F>>,
-}
-
-impl<F: RescueParameter> MembershipProofBooleanEncoding<F> {
-    fn new(nodes: &[MerkleNodeBooleanEncoding<F>]) -> Self {
-        MembershipProofBooleanEncoding {
-            nodes: nodes.to_vec(),
-        }
-    }
-}
-
-impl<F: RescueParameter> From<&MembershipProof<F>> for MembershipProofBooleanEncoding<F> {
+impl<F: RescueParameter> From<&MembershipProof<F>>
+    for MembershipProofBooleanEncoding<RescueMerkleTree<F>>
+{
```

Comment:
Maybe consider `impl<T: Borrow<MembershipProof>> From<T>` to accept conversion from both ownered and borrowed?

Reply:
Good suggestion, though I run into issues where the generic type `T` wouldn't know what is its `pos`, which is needed for a concrete impl.

I realized it's the same blocker as implementing a generic `From` for this type, which I have tried earlier to reduce code duplication. Notice how we have both:
```
impl<F: RescueParameter> From<&MembershipProof<F>>
    for MembershipProofBooleanEncoding<SparseMerkleTree<F>>
```

and 

```
impl<F: RescueParameter> From<&MembershipProof<F>>
    for MembershipProofBooleanEncoding<RescueMerkleTree<F>>
```

though of course `MembershipProof` is a different type, since both impls define their own type alias.
Maybe I'll consider adding a macro for generating this, as well as the whole impl of `MerkleTreeHelperGadget` - but we can probably leave it for another issue if I don't in this PR, it's just a nice to have.